### DISCUSSION TITLE
2D Poisson-disk sampling

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1970] [General]
Hi,

How to generate 2D Poisson-disk sampling in a specific polygon (a list of Vec2)?

Thanks.

**Comment by aismann** [comment]
> Hi,
> 
> How to generate 2D Poisson-disk sampling in a specific polygon (a list of Vec2)?
> 
> Thanks.

https://github.com/topics/poisson?l=c%2B%2B
https://github.com/thinks/poisson-disk-sampling

**Reply by paulocoutinhox** [reply]
Im talking with this guy:
https://github.com/corporateshark/poisson-disk-generator/issues/11

I just want to understand if there is any algorithm that does this more efficiently and still uses the size of the objects as a reference so as not to overwrite each other in a polygon without a specific shape.

This is my current algorithm (but not poison implementation):
![IMG_2D75A2410DD3-1](https://github.com/axmolengine/axmol/assets/395096/d712183d-1995-43f7-8e84-cc8c69b0343d)

**Reply by paulocoutinhox** [reply]
and this is my current map with polygon where is can be generated:
<img width="663" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/d183e829-3917-48fd-b19e-8d0b7632b5b8">

**Comment by smilediver** [comment]
You can try iteratively push objects from each other and from other stuff. Something like:

```
place the objects randomly
repeat n times
    for each object
        for each other_object
            diff = object.center - other_object.center
            distance = diff.length
            if distance < min_distance
                dir = diff.normalized
                push_distance = ... // either value based on how close the objects are or a constant, keep it small
                object.center += dir * push_distance
                other_object.center -= dir * push_distance
        for each wall
            ... // do same stuff
```
I'd recommend to place some big objects first, push them apart (with higher minimal distance), then add smaller objects and then push only smaller objects.
